### ISSUE TITLE
Prevent the possibility of conflicts between methods/variables provided by the framework and those defined by each module

### ISSUE THREAD

**Issue by mmcev106** [vanderbilt/redcap-external-modules#117]
This issue came out of this problem/discussion:

https://community.projectredcap.org/questions/47988/upgrade-to-870-and-fails-to-load.html

I'm thinking about moving all of the **AbstractExternalModule** methods and variables provided by the framework into a separate internal only class.  I believe this could entirely eliminate the possibility of future conflicts.  Method calls could seamlessly be forwarded to this internal instance from the normal module instance (likely via magic methods).

Does anyone see any potential problems with this approach?

@taylorr4 @mcguffk @123andy @pbchase @tbembersimeao @lsgs @swaffoja

**Comment by swaffoja** [comment]
Is this mock code about right for what you are thinking?

    class AbstractExternalModule {
        // All properties and methods from AbstractExternalModule have been moved to the
        // InternalAbstractExternaleModule class.
        private $internalAbstractExternalModule = new InternalAbstractExternalModule();
    
        public function __call($name, $args) {
            // magically passthrough method calls to InternalAbstractExternaleModule
            if (method_exists($this->internalAbstractExternalModule, $name)) {
                call_user_func_array([$this->internalAbstractExternalModule, $name], $args);
            }
        }
    }
    
    class InternalAbstractExternalModule {
        public function getConfig() {
            return ExternalModules::getConfig(...);
        }
    }
    
    class ExternalModules {
        public static function getConfig() {
            return [];
        }
    }

    class MyModule extends AbstractExternalModule {
    }

    $myModule = new MyModule();
    $config = $myModule->getConfig();


Concerns:

- IDEs would not be able to introspect/analyze the `AbstractExternalModule` class. It would disable auto-completion for both methods and properties that you want visible to developers. The following code is valid but my IDE would bark. This is a pretty big issue in my opinion. You could add DocBlocks in the `AbstractExternalModule` class so IDEs could perform introspection but then your documentation is decoupled from the actual code in `InternalAbstractExternalModule`.
 ![no_code_inspection](https://user-images.githubusercontent.com/887384/45519786-4ff3b700-b77c-11e8-9d00-1ea525ee4018.png)
- Magic method performance typically isn't as good as concrete methods. I try to avoid magic methods whenever possible aside from `__construct()`. I'd recommend testing performance as there would be a lot of magic calls.

**Comment by 123andy** [comment]
I love the simple Ide completion and echo JS’s concerns.  That said I am by no means an expert in this sort of thing.  I would rather use obscure names for newer methods than  reduce collisions, perhaps aemInit() vs init, aemLog vs log, ... for a comparable new method than risk performance issues.  I’d rather focus on usability with Mark’s precious time.

-Andy

(sent from mobile)

On Sep 13, 2018, at 4:07 PM, Jonathan Swafford <notifications@github.com<mailto:notifications@github.com>> wrote:


Is this mock code about right for what you are thinking?

class AbstractExternalModule {
    // All properties and methods from AbstractExternalModule have been moved to the
    // InternalAbstractExternaleModule class.
    private $internalAbstractExternalModule = new InternalAbstractExternalModule();

    public function __call($name, $args) {
        // magically passthrough method calls to InternalAbstractExternaleModule
        if (method_exists($this->internalAbstractExternalModule, $name)) {
            call_user_func_array([$this->internalAbstractExternalModule, $name], $args);
        }
    }
}

class InternalAbstractExternalModule {
    public function getConfig() {
        return ExternalModules::getConfig(...);
    }
}

class ExternalModules {
    public static function getConfig() {
        return [];
    }
}

class MyModule extends AbstractExternalModule {
}

$myModule = new MyModule();
$config = $myModule->getConfig();


Concerns:

  *   IDEs would not be able to introspect/analyze the AbstractExternalModule class. It would disable auto-completion for both methods and properties that you want visible to developers. The following code is valid but my IDE would bark. This is a pretty big issue in my opinion. You could add DocBlocks in the AbstractExternalModule class so IDEs could perform introspection but then your documentation is decoupled from the actual code in InternalAbstractExternalModule.
[no_code_inspection]<https://user-images.githubusercontent.com/887384/45519786-4ff3b700-b77c-11e8-9d00-1ea525ee4018.png>
  *   Magic method performance typically isn't as good as concrete methods. I try to avoid magic methods whenever possible aside from __construct(). I'd recommend testing performance as there would be a lot of magic calls.

**Comment by mmcev106** [comment]
Those are very good points.  We could only move the internal methods to a separate class (and reference them normally instead of using magic methods).  While this might be worthwhile long term, I have a more pressing concern:

I ran into an issue adding the **log()** method to **AbstractExternalModule** for the new logging functionality since the Shazam module already defined a static **log()** function.  This caused a compile time error.  Andy has since renamed the method, but I ended up using magic methods to implement the **log()** function temporarily (until I'm confident most people have upgraded Shazam to include Andy's change).

To avoid issues like this moving forward, perhaps we should discontinue support for static methods in modules.  We could start by simply displaying a big fat warning at enable time.  We don't actually ever have to disallow them unless we want to do.  I'm not as concerned about this issue for non-static methods, as overriding one unexpectedly would likely just result in the inability to use a new feature without refactoring (much better than a fatal error).  Such a warning might sound like this:

> This module defines one or more static methods.  Static methods are not supported in External Modules since they can cause conflicts with future functionality, potentially resulting in the REDCap server crashing.  We will allow this module to be enabled for now, but this may change in a future REDCap version.  To avoid future issues, please ask this module's author to refactor the module to avoid using static methods.

**Comment by mmcev106** [comment]
@taylorr4 @mcguffk @123andy @pbchase @tbembersimeao @lsgs @swaffoja

It sounds like this problem is significantly worse than I expected.  This spreadsheet documents all the possible inheritance conflict scenarios that will cause fatal errors:

https://docs.google.com/spreadsheets/d/1ReQKK7vY85QY3yuOaRWiV3UP8TU5GQi3yJvbHnEyxdg

Drupal seems to solve this problem using [traits](http://php.net/manual/en/language.oop5.traits.php), but it the larger computer science community [recommends composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance).  I am leaning towards the latter, which would necessitate some sort of secondary helper object to hold all the framework provided methods (ex: **$module->someOtherObject->allFrameworkProvidedMethods()**).  I think it might be easier to manage long term than traits.  Regardless, I think we have to act here to make sure our ecosystem can grow sustainably.  Does anyone see any other options?

I really wish I could/would have re-evaluated this part of the design of the initial External Modules prototype sooner...  There may be other looming issues like this that might warrant a full audit of the External Modules framework.

**Comment by 123andy** [comment]
I like the idea of putting all the helper methods in a sub-object...


On Sep 18, 2018, at 9:18 AM, Mark McEver <notifications@github.com<mailto:notifications@github.com>> wrote:

o

**Comment by swaffoja** [comment]
You might consider using Dependency Injection ([1](http://fabien.potencier.org/do-you-need-a-dependency-injection-container.html), [2](https://www.sitepoint.com/dependency-injection-with-pimple/)). It would let you break up your framework provided methods into smaller chunks and eliminate inheritance conflicts. It would be a significant change in development and usage though.

I like `$module->someOtherObject->allFrameworkProvidedMethods()` as it is pretty close to what you are doing now and wouldn't require a huge change. 

I also like `$module->someOtherObject->discreteConceptObject->someFrameworkMethods()`. I mention this because I've written a few large classes in my time and they become [difficult](https://dzone.com/articles/rule-30-–-when-method-class-or) to understand and maintain in my opinion. Example:

    // All framework methods in a single class (someOtherClass). Will become very 
    // big as functionality grows. Naming methods in a clear fashion difficult
    $module->someOtherObject->selectData($some, $params=[])
    $module->someOtherObject->updateData($some, $params=[])
    $module->someOtherObject->deleteData($some, $params=[])
    $module->someOtherObject->getConfig()

    // Framework methods grouped by concept
    $module->someOtherObject->data->select($some, $params=[])
    $module->someOtherObject->data->update($some, $params=[])
    $module->someOtherObject->data->delete($some, $params=[])
    $module->someOtherObject->config->get()

**Comment by mmcev106** [comment]
I that idea is definitely worth considering @swaffoja.

**Comment by taylorr4** [comment]
I think many of the ideas above make a lot of sense going forward to reduce the possibility of method/var collisions in modules, and i agree that if this continues to be a problem, it's better to solve it now than wait till later.

However, i have some major concerns here if we decide to completely change the format in which modules are written. If there is no way to allow these changes to also be backward compatible, then i think we're going to be looking at a ginormous mess that could (at worst) really sour people on external modules in the REDCap Community. For instance, if we decide to go the route of putting all the helper methods in a sub-object (which sounds fine), then that means if someone upgrades REDCap to the future version that contains this module format change, it will break all their modules unless EVERY author of EVERY module they have has put in the effort to update their modules to the new format. It seems like we'd almost have to wait until EVERY module in the REDCap Repo has been transitioned to this before we officially roll out these changes to the framework itself. Maybe that's doable, but i have a pretty good feeling that @mmcev106 and @mcguffk's team might not have the time and bandwidth to transition all of their many many modules being used. 

Also think about modules that *aren't* in the Repo that people built just for their institution. How do we let them know that these changes are coming to the framework? They might not be connected to our REDCap Community and normal communication channels to learn about this. I could imagine that many people (admins and/or developers) that may have built modules and then upgrade REDCap and then it really messes up someones multi-million dollar study that was really depending on these modules, in which they didn't know that simply upgrading REDCap would break all their existing modules. I'm just throwing out worst case scenarios here though, but these are totally plausible. 

So if we totally change the format of modules (like in the suggestions above), then if there's a quick way to add a few lines of code to make an older module compatible with the new format OR if there's a way for this to be backward compatible with older modules, that sounds awesome to me and sounds like that could really minimize the pain felt during this transition for REDCap partners. So that's my 2 cents.

**Comment by mmcev106** [comment]
I believe the sub-object solution could be implemented while maintaining full backward compatibility indefinitely.  I don't think there's any reason not to leave all existing methods in places for the foreseeable future, even if they only contain one line referencing the corresponding method under our new class(es).

**Comment by taylorr4** [comment]
@mmcev106 that definitely sounds great and makes me feel a lot better about this transition. thanks! _blood pressure going down now_

**Comment by dr01d3r** [comment]
I'm a fan of the sub-objects as well.  I use traits for my own stuff, but at a small scale.  The problem is that everything with a trait is implicit - as soon as you use it, you inherit all it's fields and functions.  With sub-objects, it is explicit, forcing you to specify the source.  I can see how that will scale better over time and become more manageable than traits.

**Comment by mmcev106** [comment]
If we wanted to make any potentially breaking changes for this, we could start versioning the AbstractExternalModule class (or something similar) so we can let module authors opt-in on their on schedule.  That might help with #148 as well.

**Comment by mmcev106** [comment]
I may have an opportunity to start on this very soon for one of my current projects.  How close does the following sound to a long term solution?

- A `framework-version` is specified in **config.json**.  We increment the `framework-version` whenever we make breaking changes.  This gives us the ability to maintain full backward compatibility indefinitely, while leaving room for any type of future changes that module authors can adopt on their own schedules.  I'm envisioning a simple scalar integer versioning scheme (1, 2, 3, etc., like [Android API Levels](https://source.android.com/setup/start/build-numbers)).  The `framework-version` would default to `1` if ommitted.  Documentation would remain available for all prior versions.

- For `"framework-version": 2`, a `framework` object is added that contains all framework provided methods as follows (old methods could still be accessed the old way):
    ```
    $module->framework->getProjectSetting(...);
    $module->framework->someNewMethod(...);
    ```
- Subobjects, dependency injection, or any other radical changes could come in any future `framework-version`.  The behavior of the `framework` object would change accordingly if/when module authors updated to that version.

**Comment by pbchase** [comment]
That seems like a good way to protect us from change while allowing you more freedom to advance the framework.

**Comment by mmcev106** [comment]
Thank you for the feedback @swaffoja & @pbchase.  I will plan to proceed as described.

**Comment by mmcev106** [comment]
May I take a thumbs-up emoji poll on the following two comments?  Which example syntax would you prefer for EM framework methods moving forward? Feel free to suggest other options as well.  These calls would take place from any method in a module class.  The specific subobjects and methods below are just possible examples.

**Comment by mmcev106** [comment]
*Option 1: An explicit "framework" object containing all other subobjects* - This would mean no refactoring would ever be required when upgrading a module to new framework versions.
```
$this->framework->records->lock(...);
$this->framework->records->unlock(...);

$this->framework->dags->create(...);
$this->framework->dags->delete(...);
```

**Comment by mmcev106** [comment]
*Option 2: Subobjects directly on the module class* - More concise, but may cause conflicts requiring module authors to perform some refactoring when upgrading to new framework versions (at a time of their choosing).
```
$this->records->lock(...);
$this->records->unlock(...);

$this->dags->create(...);
$this->dags->delete(...);
```

**Comment by taylorr4** [comment]
I'm torn. I like both for the reasons given.

**Comment by tmwil** [comment]
I think the value of no refactoring/no conflicts is well worth it in my opinion. Otherwise this will be an issue any time we want to add a sub-object.

**Comment by mmcev106** [comment]
@tmwil, I suspect you're aware of this already, but I mainly wanted to clarify for everyone else.  Every time we make a potentially breaking change going forward (like adding a subobject that might conflict) we'd increment the framework version.  Old framework versions would be fully backward compatible. 
 Module authors would update to future versions (and address potential conflicts) if/when they choose.

**Comment by mmcev106** [comment]
I have implemented the basics for option #1 and added two methods I created for the PROCLAIM project:

https://github.com/vanderbilt/redcap-external-modules/blob/testing/docs/framework/v2.md

I'm going to close this ticket, but feel free to reopen it or create additional tickets as you have more ideas.  I think this might have to mature organically over time.
